diff --git a/wolftpm/tpm2_types.h b/wolftpm/tpm2_types.h
index 6edbdee..745b17c 100644
--- a/wolftpm/tpm2_types.h
+++ b/wolftpm/tpm2_types.h
@@ -197,6 +197,14 @@ typedef int64_t  INT64;
     #define XATOI(s)          atoi((s))
 #endif
 
+/* make sure file IO macros are available for examples */
+#ifndef XFEOF /* used in pcr/extend example */
+    #define XFEOF      feof
+#endif
+#ifndef XREWIND /* used in tpm_test_keys.c */
+    #define XREWIND    rewind
+#endif
+
 /* enable way for customer to override printf */
 #ifdef XPRINTF
     #undef  printf
