From e1730c760c79b7d5ebc7613555ab8e221e7ac521 Mon Sep 17 00:00:00 2001
From: JacobBarthelmeh <jacob@wolfssl.com>
Date: Wed, 29 Mar 2023 14:11:56 -0700
Subject: [PATCH] check if colrm is available for options.h creation

---
 configure.ac | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 38afb1f1a0a..81f75b970f0 100644
--- a/configure.ac
+++ b/configure.ac
@@ -8641,8 +8641,23 @@ echo "extern \"C\" {" >> $OPTION_FILE
 echo "#endif" >> $OPTION_FILE
 echo "" >> $OPTION_FILE
 
+# check for supported command to trim option with
+which colrm &> /dev/null
+RESULT=$?
+if test "$RESULT" = "0"; then
+    TRIM="colrm 3"
+else
+    which cut &> /dev/null
+    RESULT=$?
+    if test "$RESULT" = "0"; then
+        TRIM="cut -c1-2"
+    else
+        AC_MSG_ERROR([Could not find colrm or cut to make options file])
+    fi
+fi
+
 for option in $CPPFLAGS $AM_CPPFLAGS $CFLAGS $AM_CFLAGS; do
-    opt_type=$(echo $option | colrm 3)
+    opt_type=$(echo $option | $TRIM )
     case "$opt_type" in
     -D)
         RHS_only=$(echo $option | sed 's/^-D//')
