diff --git a/scripts/build_ffi.py b/scripts/build_ffi.py
index 2e9678b..5aedfe5 100644
--- a/scripts/build_ffi.py
+++ b/scripts/build_ffi.py
@@ -375,6 +375,7 @@ def get_features(local_wolfssl, features):
     features["RSA_PSS"] = 1 if '#define WC_RSA_PSS' in defines else 0
     features["CHACHA20_POLY1305"] = 1 if '#define HAVE_CHACHA' and '#define HAVE_POLY1305' in defines else 0
     features["ML_DSA"] = 1 if '#define HAVE_DILITHIUM'  in defines else 0
+    features["ML_KEM"] = 1 if '#define HAVE_KYBER'  in defines else 0
 
     if '#define HAVE_FIPS' in defines:
         if not fips:
